### DISCUSSION TITLE
Custom validation and synchronization logic?

### DISCUSSION THREAD

**Discussion by trenta3** [appy-one/acebase#91] [Ideas]
First of all, thank you a lot for the project, which I really think is great and quite a rare thing to find.

I would like to ask whether it is currently possible to provide custom validation functions and synchronization logic: I am aware of the `.read` and `.write` attributes to `rules.json` but, apart from the difficulties in writing a file inside the database folder[^1], there are cases in which one would like to perform more complex kinds of validations on node insertion[^2], and this doesn't seem possible with current "server functions".

Moreover I think that also the ability to implement a custom synchronization logic with other databases would benefit a lot to the project, and would provide a sort-of "bring your own transport layer" like it can currently be done with storage, instead of having just the local version and the client-server one.
What I have in mind is the ability to attach custom functions which are notified of whichever modification happens in the database, so that one can stream changes to other databases. As far as I understand from reading other issues and discussions the current `mutated` event does not satisfy this condition since it can loose track of initial state of nodes, for examples on disconnections from the server.

Let me know what you think of those proposals, and have a nice weekend.

[^1]: I sincerely miss the ability to pass custom rules when creating the database in the server.
[^2]: For example to manage a chat application where users can be added or removed from the chat one would need a way to directly modify the permissions to read and write to the chat path.

**Comment by appy-one** [comment]
For additional data validation you can also use `.schema`, see [Adding schemas to enforce data rules](https://github.com/appy-one/acebase#schemas) and [Schema validation](https://github.com/appy-one/acebase-server#schema-validation). A custom `.validate` function will also be added in the future to allow checking of other data, but I don't have a timeline for that yet.

Regarding synchronization: if transaction logging is enabled on the server, any missed `mutated` and `mutations` events while offline are automatically synchronized upon reconnect using a cursor. Additionally, if you'd persist the last returned cursor from such events (check `snap.context().acebase_cursor`), you can also manually request mutations / effective changes since that cursor from the server API when your app restarts. See [Synchronization](https://github.com/appy-one/acebase-client#synchronization) in the acebase-client docs.

You can also check out the new [OpenAPI (Swagger) docs I've added to the acebase-server repo](https://github.com/appy-one/acebase-server/tree/master/src/routes) last weeks if you'd want to dive deeper into the inner workings of the `mutations` and `changes` API endpoints.

**Reply by trenta3** [reply]
Thanks for the explanations, I was aware of `.schema`. What I was talking about was custom validation through computations on data, like implementing ACL via queries on other parts of database, which I think is probably what the `.validate` function will do.

Regarding synchronizations instead this is all useful information. I will have to look more inside the docs!

**Reply by appy-one** [reply]
Implementing `.validate` has become easier now that I've ported all to TypeScript and isolated code into their own files, I'll give it a closer look after I've released all those changes in v1.10.0 which will be somewhere next week.

**Comment by trenta3** [comment]
Thank you for the work, waiting for the big release.
